File: crm-backend/schemas/email_marketing.py
"""
Schémas Pydantic pour Email Marketing & Lead Scoring

EmailIngest: Webhook pour tracking emails (Resend, etc.)
LeadScore: Score d'engagement calculé
"""

from pydantic import BaseModel, Field
from typing import Optional, List, Literal
from datetime import datetime

# Types littéraux
EmailEventType = Literal['sent', 'opened', 'clicked', 'bounced']
EmailStatus = Literal['sent', 'opened', 'clicked', 'bounced']


class EmailIngestPayload(BaseModel):
    """
    Payload reçu du webhook provider (Resend, Sendgrid, etc.)

    Permet de créer/update EmailEventTracking + calculer LeadScore
    """
    provider: str = Field(..., min_length=1, max_length=50, description="Provider name (resend, sendgrid, etc.)")
    external_id: str = Field(..., min_length=1, max_length=255, description="Provider message ID (unique)")
    event: EmailEventType = Field(..., description="Event type")
    occurred_at: datetime = Field(..., description="When the event occurred")

    # Entity (at least one required)
    person_id: Optional[int] = Field(None, gt=0, description="CRM Person ID")
    organisation_id: Optional[int] = Field(None, gt=0, description="CRM Organisation ID")

    # Optional metadata
    subject: Optional[str] = Field(None, max_length=500, description="Email subject")

    def model_post_init(self, __context):
        """Validation: au moins person_id OU organisation_id"""
        if not self.person_id and not self.organisation_id:
            raise ValueError("Au moins person_id ou organisation_id est requis")

    class Config:
        json_schema_extra = {
            "example": {
                "provider": "resend",
                "external_id": "msg_abc123xyz",
                "event": "opened",
                "occurred_at": "2025-10-24T15:30:00Z",
                "person_id": 42,
                "subject": "Nouvelle offre produit"
            }
        }


class EmailSendOut(BaseModel):
    """Réponse d'un EmailEventTracking"""
    id: int
    organisation_id: Optional[int]
    person_id: Optional[int]
    provider: str
    external_id: str
    subject: Optional[str]
    status: EmailStatus
    sent_at: Optional[datetime]
    open_count: int
    click_count: int
    last_open_at: Optional[datetime]
    last_click_at: Optional[datetime]
    interaction_id: Optional[int]
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class LeadScoreOut(BaseModel):
    """Réponse d'un LeadScore"""
    person_id: int
    score: int
    last_event_at: Optional[datetime]
    created_at: datetime
    updated_at: Optional[datetime]

    # Optional: inclure Person data (via join)
    person_name: Optional[str] = None
    person_email: Optional[str] = None

    class Config:
        from_attributes = True


class HotLeadsResponse(BaseModel):
    """Réponse de l'endpoint /marketing/leads-hot"""
    items: List[LeadScoreOut]
    threshold: int = Field(15, description="Score threshold pour Hot Lead")
    total: int

    class Config:
        json_schema_extra = {
            "example": {
                "items": [
                    {
                        "person_id": 42,
                        "score": 23,
                        "last_event_at": "2025-10-24T15:30:00Z",
                        "person_name": "Jean Dupont",
                        "person_email": "jean@example.com"
                    }
                ],
                "threshold": 15,
                "total": 8
            }
        }

File: crm-frontend/types/email-marketing.ts
/**
 * Types TypeScript pour Email Marketing & Lead Scoring
 *
 * Alignés avec le backend FastAPI:
 * - EmailSend: tracking emails (Resend, Sendgrid, etc.)
 * - LeadScore: score d'engagement calculé automatiquement
 */

import { z } from 'zod'

// ============================================
// Enums & Types
// ============================================

/**
 * Type d'événement email
 */
export type EmailEventType = 'sent' | 'opened' | 'clicked' | 'bounced'

/**
 * Statut d'un email
 */
export type EmailStatus = 'sent' | 'opened' | 'clicked' | 'bounced'

/**
 * Labels UI pour les statuts d'email
 */
export const EMAIL_STATUS_LABELS: Record<EmailStatus, string> = {
  sent: 'Envoyé',
  opened: 'Ouvert',
  clicked: 'Cliqué',
  bounced: 'Rebondi',
}

/**
 * Couleurs Tailwind par statut d'email
 */
export const EMAIL_STATUS_COLORS: Record<EmailStatus, string> = {
  sent: 'bg-gray-100 text-gray-700',
  opened: 'bg-blue-100 text-blue-700',
  clicked: 'bg-green-100 text-green-700',
  bounced: 'bg-red-100 text-red-700',
}

// ============================================
// Zod Schemas (Validation)
// ============================================

/**
 * Schéma Zod pour payload webhook email
 */
export const EmailIngestPayloadSchema = z.object({
  provider: z.string().min(1).max(50),
  external_id: z.string().min(1).max(255),
  event: z.enum(['sent', 'opened', 'clicked', 'bounced']),
  occurred_at: z.string().datetime(),
  person_id: z.number().int().positive().optional().nullable(),
  organisation_id: z.number().int().positive().optional().nullable(),
  subject: z.string().max(500).optional().nullable(),
}).refine(
  (data) => data.person_id !== undefined || data.organisation_id !== undefined,
  {
    message: 'Au moins person_id ou organisation_id est requis',
    path: ['person_id'],
  }
)

/**
 * Schéma Zod pour EmailSend (réponse API)
 */
export const EmailSendSchema = z.object({
  id: z.number().int().positive(),
  organisation_id: z.number().int().positive().optional().nullable(),
  person_id: z.number().int().positive().optional().nullable(),
  provider: z.string(),
  external_id: z.string(),
  subject: z.string().optional().nullable(),
  status: z.enum(['sent', 'opened', 'clicked', 'bounced']),
  sent_at: z.string().datetime().optional().nullable(),
  open_count: z.number().int().nonnegative(),
  click_count: z.number().int().nonnegative(),
  last_open_at: z.string().datetime().optional().nullable(),
  last_click_at: z.string().datetime().optional().nullable(),
  interaction_id: z.number().int().positive().optional().nullable(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime().optional().nullable(),
})

/**
 * Schéma Zod pour LeadScore (réponse API)
 */
export const LeadScoreSchema = z.object({
  person_id: z.number().int().positive(),
  score: z.number().int().nonnegative(),
  last_event_at: z.string().datetime().optional().nullable(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime().optional().nullable(),
  // Optional: person data (via join)
  person_name: z.string().optional().nullable(),
  person_email: z.string().optional().nullable(),
})

/**
 * Schéma Zod pour HotLeadsResponse
 */
export const HotLeadsResponseSchema = z.object({
  items: z.array(LeadScoreSchema),
  threshold: z.number().int().nonnegative(),
  total: z.number().int().nonnegative(),
})

// ============================================
// TypeScript Interfaces
// ============================================

/**
 * Payload webhook email
 */
export interface EmailIngestPayload {
  provider: string
  external_id: string
  event: EmailEventType
  occurred_at: string  // ISO datetime
  person_id?: number | null
  organisation_id?: number | null
  subject?: string | null
}

/**
 * EmailSend (réponse API)
 */
export interface EmailSend {
  id: number
  organisation_id?: number | null
  person_id?: number | null
  provider: string
  external_id: string
  subject?: string | null
  status: EmailStatus
  sent_at?: string | null  // ISO datetime
  open_count: number
  click_count: number
  last_open_at?: string | null  // ISO datetime
  last_click_at?: string | null  // ISO datetime
  interaction_id?: number | null
  created_at: string  // ISO datetime
  updated_at?: string | null  // ISO datetime
}

/**
 * LeadScore (réponse API)
 */
export interface LeadScore {
  person_id: number
  score: number
  last_event_at?: string | null  // ISO datetime
  created_at: string  // ISO datetime
  updated_at?: string | null  // ISO datetime
  // Optional: person data (via join)
  person_name?: string | null
  person_email?: string | null
}

/**
 * Réponse endpoint /marketing/leads-hot
 */
export interface HotLeadsResponse {
  items: LeadScore[]
  threshold: number
  total: number
}
